mono/utils/mono-threads.c

```diff
@@ -1835,3 +1911,181 @@ mono_thread_info_get_tools_data (void)
 
 	return info ? info->tools_data : NULL;
 }
+
+typedef struct {
+	MonoLockFreeQueueNode node;
+
+	// For cleanup metadata
+	MonoUtilityThread *thread;
+
+	// For synch calls
+	gboolean *finished;
+	MonoSemType *response_sem;
+
+	// Variably-sized, size is thread->payload_size
+	gpointer payload [MONO_ZERO_LEN_ARRAY]; 
+} UtilityThreadQueueEntry;
+
+static void
+free_queue_entry (gpointer p)
+{
+	UtilityThreadQueueEntry *util = (UtilityThreadQueueEntry *) p;
+	mono_lock_free_free (p, util->thread->message_block_size);
+}
+
+static gboolean
+utility_thread_handle_inbox (MonoUtilityThread *thread)
+{
+	UtilityThreadQueueEntry *entry = (UtilityThreadQueueEntry *) mono_lock_free_queue_dequeue (&thread->work_queue);
+	if (!entry)
+		return FALSE;
+	
+	thread->callbacks.command (thread->state_ptr, &entry->payload);
+	if (entry->response_sem) {
+		*entry->finished = TRUE;
+		mono_os_sem_post (entry->response_sem);
+	}
+
+	mono_thread_hazardous_try_free (entry, free_queue_entry);
+
+	return TRUE;
+}
+
+static void *
+utility_thread (void *arg)
+{
+	mono_thread_info_wait_inited ();
+	mono_thread_info_attach ();
```

Comment:
For the profiler to use this API, there needs to also be an early init callback that's invoked before `mono_thread_info_attach` is called.

Reply:
I wonder if it would be safe to just move the init call up and to say that you can't use certain primitives in the init function

Reply:
The issue is that the profiler needs to do stuff both before and after `mono_thread_info_attach` is called, during utility thread init.

Although, I don't recall if it is safe to call `mono_thread_info_attach` on an already-attached thread, but if it is, I suppose moving the init callback earlier and having the profiler just call `mono_thread_info_attach` as appropriate in its callback would be an OK solution too.